libkbfs/md_journal.go

```diff
@@ -0,0 +1,507 @@
+	return j.checkGetParams(currentUID)
+}
+
+func (j mdJournal) getRange(
+	currentUID keybase1.UID, start, stop MetadataRevision) (
+	[]ImmutableBareRootMetadata, error) {
+	_, err := j.checkGetParams(currentUID)
+	if err != nil {
+		return nil, err
+	}
+
+	realStart, mdIDs, err := j.j.getRange(start, stop)
+	if err != nil {
+		return nil, err
+	}
+	var rmds []ImmutableBareRootMetadata
+	for i, mdID := range mdIDs {
+		expectedRevision := realStart + MetadataRevision(i)
+		rmd, err := j.getMD(mdID)
+		if err != nil {
+			return nil, err
+		}
+		if expectedRevision != rmd.Revision {
+			panic(fmt.Errorf("expected revision %v, got %v",
+				expectedRevision, rmd.Revision))
+		}
+		irmd := MakeImmutableBareRootMetadata(rmd, mdID)
+		rmds = append(rmds, irmd)
+	}
+
+	return rmds, nil
+}
+
+// MDJournalConflictError is an error that is returned when a put
+// detects a rewritten journal.
+type MDJournalConflictError struct{}
+
+func (e MDJournalConflictError) Error() string {
+	return "MD journal conflict error"
+}
+
+func (j mdJournal) put(
+	ctx context.Context, signer cryptoSigner, ekg encryptionKeyGetter,
+	rmd *RootMetadata, currentUID keybase1.UID,
+	currentVerifyingKey VerifyingKey) (MdID, error) {
+	head, err := j.getHeadHelper()
+	if err != nil {
+		return MdID{}, err
+	}
+
+	mStatus := rmd.MergedStatus()
+	bid := rmd.BID
+
+	if (mStatus == Unmerged) && (bid == NullBranchID) &&
+		(head != ImmutableBareRootMetadata{}) {
+		rmd.BID = head.BID
+		rmd.PrevRoot = head.mdID
+		bid = rmd.BID
+	}
+
+	if (mStatus == Merged) != (bid == NullBranchID) {
+		return MdID{}, errors.New("Invalid branch ID")
+	}
+
+	// Check permissions and consistency with head, if it exists.
+	if head != (ImmutableBareRootMetadata{}) {
+		ok, err := isWriterOrValidRekey(
+			j.codec, currentUID, head.BareRootMetadata,
+			&rmd.BareRootMetadata)
+		if err != nil {
+			return MdID{}, err
+		}
+		if !ok {
+			// TODO: Use a non-server error.
+			return MdID{}, MDServerErrorUnauthorized{}
+		}
+
+		// If we're trying to push a merged MD onto a branch,
+		// return a conflict error so the caller can retry
+		// with an unmerged MD.
+		if mStatus == Merged && head.BID != NullBranchID {
+			return MdID{}, MDJournalConflictError{}
+		}
+
+		// Consistency checks
+		err = head.CheckValidSuccessorForServer(
+			head.mdID, &rmd.BareRootMetadata)
+		if err != nil {
+			return MdID{}, err
+		}
+	}
+
+	var brmd BareRootMetadata
+	err = encryptMDPrivateData(
+		ctx, j.codec, j.crypto, signer, ekg,
+		currentUID, rmd.ReadOnly(), &brmd)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	id, err := j.putMD(&brmd, currentUID, currentVerifyingKey)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	err = j.j.append(brmd.Revision, id)
+	if err != nil {
+		return MdID{}, err
+	}
+
+	return id, nil
+}
+
+// flushOne sends the earliest MD in the journal to the given MDServer
+// if one exists, and then removes it. Returns whether there was an MD
+// that was put.
+func (j mdJournal) flushOne(
```

Comment:
Added a comment for `mdJournal` that it isn't goroutine-safe. (All the other functions rely on this, too.)